rfcs/records.md

```diff
@@ -0,0 +1,448 @@
+The field lookup data is internal and immutable; for example, it might contain a string->index dictionary to be able to quickly locate fields in internal storage.
+
+Reading and writing fields from a record uses `.` or `[]` operator; like tables, passing the field name returns the field value. Unlike tables, if the field is not
+present in the table, the error is raised. This is in constrast with tables where `nil` is returned for unknown keys upon read; records are meant to be stricted than
+tables and as such returning `nil` will mask valuable errors, and make it more difficult to be strict about the types of the result.
+
+The field lookup does not use `__index` or `__newindex` metamethods (or the metatable in general).
+
+Invoking methods with `:` desugars into `getmetatable(obj).method(obj, args)` instead of the usual `obj.method(obj, args)`. This is important because it allows to
+keep the method calls as efficient as possible, as they don't need to check whether the record has a given method as a field.
+
+> TODO: Should we use `__namecall` instead of raw MT access? It seems more consistent, but at the same time `__namecall` today expects a function so it might be best
+> to leave it as is?
+
+> TODO: How do we expose the record keys? Should it be a builtin? Accessible through shape? Not available initially?
+
+### Defining and constructing records
+
+To define a record, you need to create the shape, which you can do using the newly introduced syntax with a context-sensitive `record` keyword.
+
+> TODO: The draft RFC suggests two options for the syntax; only one will be chosen in the final version
+
+Syntax A:
+
+```lua
+record Person = { name: string, age: number }
+
+-- types can be omitted and default to any
+record Point = { x, y }
+```
+
+Syntax B:
+
+```lua
+record Person(name: string, age: number)
+-- types can be omitted and default to any
+record Point(x, y)
+```
+
+This defines `Point` simultaneously as a local variable that corresponds to the shape table, and a type variable that corresponds to the record type.
+
+The resulting shape table automatically is set up to be a valid record shape, but can still be modified by adding methods to it:
+
+```lua
+function Point.__add(l, r)
+    return Point(l.x + r.x, l.y + r.y)
+end
+
+function Point:sum()
+    return self.x + self.y
+end
+
+function Point.newDiagonal(v)
+    return Point(v, v)
+end
+```
+
+Note that `Point` is simply a table and as such it can be used to store static methods as well; as it also serves as a metatable, metafields defined on this table
+will change the behavior of the record values. The shape isn't frozen automatically but can be frozen manually if desired via `table.freeze`.
+
+To create a record, you need to use a record constructor. This is done using call-like syntax:
+
+Syntax A:
+
+```lua
+local person = Person { name = "Bob", age = 42 }
```

Comment:
There's optionality here - we could go either way.

One option is that Person { fields } is special magic syntax (that's backwards compatible with function call), and Person isn't callable otherwise.
Another option is that Person { fields } semantically behaves just as it does today, and the fields are copied out of the source table into the record - and compiler does some magic to optimize this to avoid a temporary table object.

In either case you'd get a fresh object.

Reply:
Got it, a temporary borrow.

I think the latter option is going to be better. It's going to be able to cooperate with the rest of the existing ecosystem. If `map` applies the function `f` to each element in the array `a` in the form `f(v)`, and we have some record `City`, then calling `map(cities, City)` is an error. It doesn't quite make much sense for a feature to be partially exclusive to other language features.